adapters/repos/db/shard_shutdown.go

```diff
@@ -45,7 +51,29 @@ import (
 // idempotent Shutdown methods. In other parts, it explicitly checks if a
 // component was initialized. If not, it turns it into a noop to prevent
 // blocking.
-func (s *Shard) Shutdown(ctx context.Context) (err error) {
+func (s *Shard) performShutdown(ctx context.Context) (err error) {
+	s.shutdownLock.Lock()
+	defer s.shutdownLock.Unlock()
```

Comment:
this lock is kept for entire shutdown execution. meaning any request incoming while shard is being shut will have to wait until method is complete (`s.shutdownLock.Lock()` in `preventShutdown`), which may result in timeouts. Is that expected?

Reply:
yes, but I added a reject anyway